File: src/Systems/C++/example3.cpp
#include <iostream>

// Abstract Product
class Report {
public:
    virtual void generate() = 0;
};

// Concrete Product
class PDFReport : public Report {
public:
    void generate() override {
        std::cout << "Generating PDF report" << std::endl;
    }
};

class HTMLReport : public Report {
public:
    void generate() override {
        std::cout << "Generating HTML report" << std::endl;
    }
};

// Abstract Factory
class ReportFactory {
public:
    virtual Report* createReport() = 0;
};

// Concrete Factory
class PDFReportFactory : public ReportFactory {
public:
    Report* createReport() override {
        return new PDFReport();
    }
};

class HTMLReportFactory : public ReportFactory {
public:
    Report* createReport() override {
        return new HTMLReport();
    }
};

// Usage
void useFactory(ReportFactory* factory) {
    Report* report = factory->createReport();
    report->generate();
    delete report;
}

int main() {
    ReportFactory* pdfFactory = new PDFReportFactory();
    ReportFactory* htmlFactory = new HTMLReportFactory();
    useFactory(pdfFactory);
    useFactory(htmlFactory);
    delete pdfFactory;
    delete htmlFactory;
    return 0;
}

File: src/Systems/Go/example3.go
package main

import "fmt"

// Abstract Product
type Report interface {
    Generate()
}

// Concrete Product
type PDFReport struct{}

func (p *PDFReport) Generate() {
    fmt.Println("Generating PDF report")
}

type HTMLReport struct{}

func (h *HTMLReport) Generate() {
    fmt.Println("Generating HTML report")
}

// Abstract Factory
type ReportFactory interface {
    CreateReport() Report
}

// Concrete Factory
type PDFReportFactory struct{}

func (p *PDFReportFactory) CreateReport() Report {
    return &PDFReport{}
}

type HTMLReportFactory struct{}

func (h *HTMLReportFactory) CreateReport() Report {
    return &HTMLReport{}
}

func useFactory(factory ReportFactory) {
    report := factory.CreateReport()
    report.Generate()
}

func main() {
    pdfFactory := &PDFReportFactory{}
    htmlFactory := &HTMLReportFactory{}
    useFactory(pdfFactory)
    useFactory(htmlFactory)
}
